netket/optimizer/stochastic_reconfiguration.py

```diff
@@ -0,0 +1,303 @@
+                    self._S,
+                    grad,
+                    cond=self._svd_threshold,
+                    lapack_driver=self._lapack_driver,
+                )
+
+                self._revert_preconditioning(out)
+
+        else:
+            if self._use_iterative:
+                op = self._linear_operator(oks, n_samp)
+
+                if self._x0 is None:
+                    self._x0 = _np.zeros(n_par)
+
+                out.real, info = self._sparse_solver(
+                    op,
+                    grad.real,
+                    x0=self._x0,
+                    tol=self.sparse_tol,
+                    maxiter=self.sparse_maxiter,
+                )
+                if info < 0:
+                    raise RuntimeError("SR sparse solver did not converge.")
+                self._x0 = out.real
+            else:
+                self._S = _np.matmul(oks.conj().T, oks, self._S)
+                self._S /= float(n_samp)
+
+                self._apply_preconditioning(grad)
+
+                out.real, residuals, self._last_rank, s_vals = _lstsq(
+                    self._S.real,
+                    grad.real,
+                    cond=self._svd_threshold,
+                    lapack_driver=self._lapack_driver,
+                )
+
+                self._revert_preconditioning(out.real)
+
+            out.imag.fill(0.0)
+        self._comm.bcast(out, root=0)
+        self._comm.barrier()
+        return out
+
+    def _apply_preconditioning(self, grad):
+        if self._scale_invariant_pc:
+            # Even if S is complex, its diagonal elements should be real since it
+            # is Hermitian.
+            self._diag_S = _np.sqrt(self._S.diagonal().real)
+
+            cutoff = 1.0e-10
+
+            index = self._diag_S <= cutoff
+            self._diag_S[index] = 1.0
+            self._S[index, :].fill(0.0)
+            self._S[:, index].fill(0.0)
+            _np.fill_diagonal(self._S, 1.0)
+
+            self._S /= _np.vdot(self._diag_S, self._diag_S)
+
+            grad /= self._diag_S
+
+        # Apply diagonal shift
+        self._S += self._diag_shift * _np.eye(self._S.shape[0])
+
+    @property
+    def last_rank(self):
+        return self._last_rank
+
+    @property
+    def last_covariance_matrix(self):
+        return self._S
+
+    def _revert_preconditioning(self, out):
+        if self._scale_invariant_pc:
+            out /= self._diag_S
+
+    @property
+    def scale_invariant_regularization(self):
+        r"""bool: Whether to use the scale-invariant regularization as described by
+                    Becca and Sorella (2017), pp. 143-144.
+                    https://doi.org/10.1017/9781316417041
+        """
+        return self._scale_invariant_pc
+
+    @scale_invariant_regularization.setter
+    def scale_invariant_regularization(self, activate):
+        assert activate is True or activate is False
+        self._scale_invariant_pc = activate
```

Comment:
This line should be below the `if` block, so that the internal state of `sr` is not changed when the exception is thrown. (Otherwise, it could be confusing when someone uses SR interactively.)